### ISSUE TITLE
[bug] TypeScript non-null assertion operator `!` parsed with wrong precedence

### ISSUE THREAD

**Issue by andersk** [ast-grep/ast-grep#1184]
The TypeScript compiler parses the [non-null assertion operator `!`](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) with precedence higher than `new`, but ast-grep incorrectly parses it with precedence lower than `||` and `??`.

# ⏯ Playground Link

[Playground link with relevant code](https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoidHlwZXNjcmlwdCIsInF1ZXJ5IjoiJFghIiwicmV3cml0ZSI6Im51bGx0aHJvd3MoJFgpIiwiY29uZmlnIjoiIyBZQU1MIFJ1bGUgaXMgbW9yZSBwb3dlcmZ1bCFcbiMgaHR0cHM6Ly9hc3QtZ3JlcC5naXRodWIuaW8vZ3VpZGUvcnVsZS1jb25maWcuaHRtbCNydWxlXG5ydWxlOlxuICBhbnk6XG4gICAgLSBwYXR0ZXJuOiBjb25zb2xlLmxvZygkQSlcbiAgICAtIHBhdHRlcm46IGNvbnNvbGUuZGVidWcoJEEpXG5maXg6XG4gIGxvZ2dlci5sb2coJEEpIiwic291cmNlIjoiY29uc3QgYyA9IGEgJiYgYiE7In0=)

# 💻 Code

Code:

```ts
const c = a && b!;
```

Pattern:

```
$X!
```

# 🙁 Actual behavior

Matches `a && b!`, with `$X` bound to `a && b`.

# 🙂 Expected behavior

Matches `b!`, with `$X` bound to `b`.

**Comment by andersk** [comment]
Presumably caused by

- tree-sitter/tree-sitter-typescript#299

**Comment by HerringtonDarkholme** [comment]
Indeed! thanks for the report. I will make an update after tree-sitter upstream has fixed this.

**Comment by HerringtonDarkholme** [comment]
unfortunately it is not fixable by ast-grep :/
